plugins/chat/app/services/chat/auto_join_channels.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+module Chat
+  class AutoJoinChannels
+    include Service::Base
+
+    ALLOWED_GROUP_PERMISSIONS ||= [
+      CategoryGroup.permission_types[:create_post],
+      CategoryGroup.permission_types[:full],
+    ]
+
+    policy :chat_enabled?
+
+    contract do
+      attribute :user_id, :integer
+      attribute :channel_id, :integer
+      attribute :category_id, :integer
+    end
+
+    step :create_memberships
+
+    private
+
+    def chat_enabled?
+      SiteSetting.chat_enabled
+    end
+
+    def create_memberships
+      automatic = ::Chat::UserChatChannelMembership.join_modes[:automatic]
+      group_permissions = ALLOWED_GROUP_PERMISSIONS
+      group_ids = SiteSetting.chat_allowed_groups_map
+      everyone_allowed = group_ids.include?(Group::AUTO_GROUPS[:everyone])
+      max_users = SiteSetting.max_chat_auto_joined_users
+      now = Time.zone.now
+      last_seen_at = 30.days.ago
+
+      # used to filter down to a specific user, chat channel, or category
+      user_sql = context.user_id.to_i > 0 ? "AND u.id = #{context.user_id}" : ""
+      channel_sql = context.channel_id.to_i > 0 ? "AND cc.id = #{context.channel_id}" : ""
+      category_sql = context.category_id.to_i > 0 ? "AND c.id = #{context.category_id}" : ""
+
+      sql = <<~SQL
+        WITH chat_users AS ( -- users that are allowed to join chat
+          SELECT u.id
+            FROM users u
+            JOIN user_options uo ON uo.user_id = u.id AND uo.chat_enabled = TRUE
+           WHERE u.id > 0
+             AND u.active = TRUE
+             AND u.staged = FALSE
+             AND (u.suspended_till IS NULL OR u.suspended_till <= :now)
+             AND (u.silenced_till IS NULL OR u.silenced_till <= :now)
+             AND NOT EXISTS (SELECT 1 FROM anonymous_users au WHERE au.user_id = u.id)
+             AND u.last_seen_at > :last_seen_at
+             #{user_sql}
+             #{everyone_allowed ? "" : "AND EXISTS (SELECT 1 FROM group_users gu WHERE gu.user_id = u.id AND gu.group_id IN (:group_ids))"}
+           ORDER BY u.last_seen_at DESC
+           LIMIT :max_users
+        ), valid_chat_channels AS ( -- auto joinable chat channels
+          SELECT cc.id, cc.chatable_id
+            FROM chat_channels cc
+           WHERE cc.auto_join_users = TRUE
+             AND cc.chatable_type = 'Category'
+             AND cc.deleted_at IS NULL
+             AND cc.user_count < :max_users
```

Comment:
If a chat channel has more than `SiteSetting.max_chat_auto_joined_users` users, we no longer allow it to be auto-joined. Just confirming that this is indeed the behaviour we want?

Reply:
That's my understanding on the setting, but will as on dev to confirm.

Reply:
That's indeed the confirmed behavior.